prestodb/presto
# Optimize spark split enumeration

When enumerating splits for large tables holding all the splits inmemory
can potentially be very expensive. In our deployment a single HiveSplit
takes 1.5kb of memory. However multiple splits are extremely
compressible. This patch introduces iterative split enumeration. Splits
are loaded and serialized+compressed in batches.

Test plan:

- Implement and run extensive unit tests to make sure split enumeration logic is not broken when enumerating in batches
- Add an in memory integration test with a batch size set to `1`, `2` and `4` splits to make sure the code path is exercised
- Run problematic production query that used to trigger driver OOM to make sure the patch resolves the problem
- Run standard verifier suite with standard settings as well as with altered batch size to `1`, `2` and `5` splits

```
== NO RELEASE NOTE ==
```
